android_sample/app/src/main/jni/amber_script.cpp

```diff
@@ -0,0 +1,144 @@
+  size_t shader_id_pos =
+      shader_name.find_last_of('.', spv_extension_pos - 1UL) + 1UL;
+  return shader_name.substr(shader_id_pos, spv_extension_pos - shader_id_pos);
+}
+
+}  // namespace
+
+AmberScriptLoader::AmberScriptLoader(android_app* app) : app_context_(app) {}
+
+AmberScriptLoader::~AmberScriptLoader() = default;
+
+Result AmberScriptLoader::LoadAllScriptsFromAsset() {
+  FindAllScripts();
+  if (script_info_.empty())
+    return Result("No Amber script found");
+
+  for (auto& info : script_info_) {
+    info.script_content = ReadScript(info.asset_name);
+    if (info.script_content.empty())
+      return Result(info.asset_name + ":\n\tEmpty Amber script");
+  }
+
+  for (auto& info : script_info_) {
+    auto shader_names = GetShaderNamesForAmberScript(info.asset_name);
+    for (const auto& shader : shader_names) {
+      auto shader_content = ReadAssetContentAsVectorUint32(shader);
+      if (shader_content.empty())
+        return Result(shader + ":\n\tEmpty shader");
+
+      info.shader_map[GetShaderID(shader)] = shader_content;
+    }
+  }
+
+  return {};
+}
+
+void AmberScriptLoader::FindAllScripts() {
+  AAssetDir* asset = AAssetManager_openDir(app_context_->activity->assetManager,
+                                           kAmberDir.c_str());
+  for (const char* file_name = AAssetDir_getNextFileName(asset); file_name;
+       file_name = AAssetDir_getNextFileName(asset)) {
+    std::string file_name_in_string(file_name);
+    if (IsEndedWith(file_name_in_string, kAmberScriptExtension)) {
+      script_info_.emplace_back();
+      script_info_.back().asset_name = file_name_in_string;
+    }
+  }
+  AAssetDir_close(asset);
+}
+
+std::vector<std::string> AmberScriptLoader::GetShaderNamesForAmberScript(
+    const std::string& script_name) {
+  std::vector<std::string> shaders;
+
+  AAssetDir* asset = AAssetManager_openDir(app_context_->activity->assetManager,
+                                           kAmberDir.c_str());
+  for (const char* file_name = AAssetDir_getNextFileName(asset); file_name;
+       file_name = AAssetDir_getNextFileName(asset)) {
+    std::string file_name_in_string(file_name);
+    if (IsStartedWith(file_name_in_string,
+                      script_name + kShaderNameSignature) &&
+        IsEndedWith(file_name_in_string, kShaderExtension)) {
+      shaders.push_back(file_name_in_string);
+    }
+  }
+  AAssetDir_close(asset);
+
+  return shaders;
+}
+
+std::vector<uint32_t> AmberScriptLoader::ReadAssetContentAsVectorUint32(
+    const std::string& asset_name) {
+  auto asset_path = kAmberDir + asset_name;
+  AAsset* asset = AAssetManager_open(app_context_->activity->assetManager,
+                                     asset_path.c_str(), AASSET_MODE_BUFFER);
+  if (!asset)
+    return std::vector<uint32_t>();
+
+  size_t size_in_bytes = AAsset_getLength(asset);
+  // Initialize vector as zeros.
+  std::vector<uint32_t> content((size_in_bytes + sizeof(uint32_t) - 1) /
```

Comment:
Why +sizeof(uint32_t) - 1?

Reply:
I don't think that does that though, it will move it up to one size larger.  This seems like we should be reading into a std::vector<uint8_t> as size is in bytes. For amber script files that's what we want anyway, as it isn't uint32_t data. For spv files, we can move into a uint32_t vector afterwards.

Reply:
done